RecoEcal/EgammaCoreTools/src/EcalClusterLazyTools.cc

```diff
@@ -22,76 +22,56 @@
 #include "RecoLocalCalo/EcalRecAlgos/interface/EcalSeverityLevelAlgoRcd.h"
 #include "RecoLocalCalo/EcalRecAlgos/interface/EcalSeverityLevelAlgo.h"
 
-EcalClusterLazyToolsBase::EcalClusterLazyToolsBase( const edm::Event &ev, const edm::EventSetup &es, edm::EDGetTokenT<EcalRecHitCollection> token1, edm::EDGetTokenT<EcalRecHitCollection> token2) {
-
-  ebRHToken_ = token1;
-  eeRHToken_ = token2;
- 
-  getGeometry( es , false );
-  getTopology( es );
-  getEBRecHits( ev );
-  getEERecHits( ev );
+namespace {
+
+  template<class T, class R>
+  auto getFromEventSetup(edm::EventSetup const& eventSetup) {
+    edm::ESHandle<T> handle;
+    eventSetup.get<R>().get(handle);
+    return handle.product();
+  }
+
```

Comment:
@slava77, so this function (and its event friend) is defined now repeatedly throughout CMSSW like this (I have my own version which returns a handle). While its a small simple function and is hard to screw up, part of me recoils at such copying/pasting.   Is there a central helper function like this we all could just use? If now, could we make one?

Reply:
@Dr15Jones @makortel 
what do you think of having this somewhere in FWCore?
The use case here is of  a  single-call pattern as in a constructor initialization list.

Reply:
indeed, it simplifies things. Another example (and I'm sure there are many) is
https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaElectronProducers/plugins/ElectronNHitSeedProducer.cc#L67

Reply:
So you want the ability to call a function and get the ES product by return value or throw if missing? If so, then I'd propose adding an overload to `edm::EventSetupRecord::get` as
```
template<typename T>
T const& get() const nothrow(false);

template<typename T>
T const& get(ESInputTag const&) const nothrow(false);

template<typename T>
T const& get(ESGetTokenT<T> const&) const nothrow(false);
```
and equivalent ones which do not throw if missing
```
template<typename T>
T const* getNoThrow() const;
...
```
(Note, this can actually throw if an exception happened while the data product was being produced, it will not throw if the data product is just not present.)

Then to do what you want you'd call
```
auto const& foo = iEventSetup.get<FooRcd>().get<Foo>();
```

[I was prefering the return of a `const &` since it is absolutely guaranteed to not be nullptr.]

Reply:
To me that sounds very good @Dr15Jones , I like the idea with the overloads for pointers and references! And that means we'd get the product directly, right? @Sam-Harper mentioned before that his version of that function returns a handle, maybe there are cases where the handle is needed? I can't think of any thought.

Reply:
Yes, you'd get the product directly. As for the handle, there isn't really much use to it for the EventSetup system.